[Fix #3148] Make Rails/UniqBeforePluck more conservative

tjwp opened the pull request:
This change is to address Issue https://github.com/bbatsov/rubocop/issues/3148

Autocorrect is disabled by default for this cop since it can potentially generate false positives.

Add offenses for `distinct` as well as `uniq`.

By default only add offenses for calls to `pluck` directly
on constants.

Add an 'aggressive' mode that adds all offenses as the cop did
originally.

* [x] Wrote [good commit messages][1].
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it)
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

jonas054 commented on spec/rubocop/cop/rails/uniq_before_pluck_spec.rb:
```diff
@@ -3,55 +3,83 @@
 
-  shared_examples_for 'UniqBeforePluck cop' do |source, action|
+  shared_examples_for 'UniqBeforePluck cop' do |method, source, action|
     if action == :correct
```
The action should be called `:report` if we don't auto-correct. But I wonder if it's really necessary to remove auto-correct now that we have a conservative mode. I said in #3148 that it might be necessary to remove it, but then again, it might not be necessary. :smile: What do you think?

tjwp replied:
I'm leaning towards restoring auto-correct, but making it disabled by default. Even with the conservative mode, I think there is a possibility of false positives because the cop only checks that `pluck` is called on a constant but it cannot verify that the constant refers to a model class.

jonas054 replied:
That's reasonable. It's difficult to guess what risk auto-correct poses in this case, so go with that idea.

tjwp commented:
@jonas054 Restored auto-correct, including support for `distinct` and `uniq`, squashed, and rebased. Auto-correct is disabled by default for this cop.

jonas054 commented on lib/rubocop/cop/rails/uniq_before_pluck.rb:
```diff
@@ -4,61 +4,83 @@
+      #   # an association on an instance will return a CollectionProxy
+      #   instance.assoc.pluck(:id).uniq
+      #
+      # Autocorrect is disabled by default for this cop since it may generated
```
generated -> generate

tjwp replied:
Fixed typo.

jonas054 commented:
:+1: Changes look good. Just one more thing. Sorry I didn't mention it earlier. If this fixes #3148, which I think it does, you should start the commit message with `[Fix #3148]`.

tjwp commented:
@jonas054 Updated the commit message to start with the issue being fixed. Thanks!

bbatsov commented:
You'll have to rebase on top of the current `master`.

tjwp commented:
@bbatsov rebased
